kernel/comps/keyboard/src/i8042_chip/mouse.rs

```diff
@@ -0,0 +1,472 @@
+    }
+}
+
+/// Mouse packet state machine for handling 3-byte or 4-byte PS/2 mouse packets.
+struct MousePacketState {
+    state: PacketState,
+    packet: [u8; PACKET_LEN_INTELLIMOUSE],
+    byte_index: usize,
+    mouse_type: MouseType,
+    prev_left: AtomicBool,
+    prev_right: AtomicBool,
+    prev_middle: AtomicBool,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum PacketState {
+    WaitingForFirstByte,
+    CollectingPacket,
+}
+
+impl MousePacketState {
+    const fn new() -> Self {
+        Self {
+            state: PacketState::WaitingForFirstByte,
+            packet: [0; PACKET_LEN_INTELLIMOUSE],
+            byte_index: 0,
+            mouse_type: MouseType::Standard,
+            prev_left: AtomicBool::new(false),
+            prev_right: AtomicBool::new(false),
+            prev_middle: AtomicBool::new(false),
+        }
+    }
+
+    fn set_mouse_type(&mut self, mouse_type: MouseType) {
+        self.mouse_type = mouse_type;
+        self.reset();
+    }
+
+    /// Processes a mouse data byte and returns input events if a complete packet is ready.
+    fn process_byte(&mut self, data: u8) -> Option<Vec<InputEvent>> {
+        const PACKET_SYNC_MASK: u8 = 0x08;
+
+        match self.state {
+            PacketState::WaitingForFirstByte => {
+                // First byte must have bit 3 set (always 1 in PS/2 mouse packets).
+                if data & PACKET_SYNC_MASK != 0 {
+                    self.packet[0] = data;
+                    self.byte_index = 1;
+                    self.state = PacketState::CollectingPacket;
+                }
+                None
+            }
+            PacketState::CollectingPacket => {
+                self.packet[self.byte_index] = data;
+                self.byte_index += 1;
+
+                if self.byte_index >= self.mouse_type.packet_len() {
+                    // Complete packet received.
+                    let events = self.parse_packet();
+                    self.reset();
+                    Some(events)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// Parses a complete PS/2 mouse packet and returns input events.
+    fn parse_packet(&mut self) -> Vec<InputEvent> {
+        const BUTTON_LEFT_MASK: u8 = 0x01;
+        const BUTTON_RIGHT_MASK: u8 = 0x02;
+        const BUTTON_MIDDLE_MASK: u8 = 0x04;
+        const INDEX_STATUS: usize = 0;
+        const INDEX_X: usize = 1;
+        const INDEX_Y: usize = 2;
+        const INDEX_WHEEL: usize = 3;
+
+        let mut events = Vec::new();
+
+        let status = self.packet[INDEX_STATUS];
+        let x_delta = self.packet[INDEX_X] as i8;
+        let y_delta = self.packet[INDEX_Y] as i8;
+
+        let left_button = (status & BUTTON_LEFT_MASK) != 0;
+        let right_button = (status & BUTTON_RIGHT_MASK) != 0;
+        let middle_button = (status & BUTTON_MIDDLE_MASK) != 0;
+
+        let prev_left = self.prev_left.swap(left_button, Ordering::Relaxed);
```

Comment:
No need to use `Atomic*` variables since we have `&mut self` here. Use `prev_left: bool` directly instead.